energietransitie/twomes-backoffice-api
# Create admin CLI

## What does this change

- New Admin CLI.
- Admin CLI is **ONLY** accessible by using docker exec, which requires permission to portainer _or_ the server.
- Admin CLI can list, create, delete, reactivate and change expiry of admins.
- Admins have a separate sqlite database, which is just a file in the same directory/volume as the signing keys.
- Admin tokens need to match an admin entry in the database.
- Admin tokens can be invalidated by reactivating an admin with the CLI.
- Admin accounts can now expire, allowing creation of temporary admins (useful for students)

The easiest way to use this is on the command line of the server. Use the following command for the test container to get help for the Admin CLI:
```bash
docker exec -i twomes-api-v2-tst admin-cli help
```

## Testing

- [ ] This change does not need testing.
- [x] I tested this change.
- [ ] I did not test this change.


To test this change locally:
1. Checkout this branch.
2. Run a local test instance:
    ```bash
    docker-compose up --build
    ```
3. Run this command in a separate terminal:
    ```bash
    docker exec -i twomes-backoffice-api-local-web admin-cli help
    ```

4. Try the different commands from the cli listed in the help menu.
5. Use the printed tokens as bearer auth in postman requests to for example `http://localhost:8080/app`.
6. Try to use the same token, not the new one after reactivating an admin. You should see the token does not work anymore.
7. To fully remove all the local dev containers after testing, run this command:
    ```bash
    docker-compose down --remove-orphans
    ```

## Trello card(s)

Deploy API v2 test: https://trello.com/c/6EcMniIY
